securedrop/bin/dev-shell

```diff
@@ -6,53 +6,91 @@
 
 if ! test -n "${PYTHON_VERSION:-}" ; then
     PYTHON_VERSION=3
 fi
 
-function validate_python_version() {
+validate_python_version() {
     if [[ "$2" != "2" && "$2" != "3" ]]
     then
         echo "PYTHON_VERSION must be 2 or 3"
         exit 1
     fi
 }
 
-function docker_image() {
+## Get a short identifier for the working directory
+hashpath() {
+    pwd | sum | awk '{print $1}'
+}
+
+## Get an integer prefix for exposed ports, to support running dev and
+## test, or containers from different directories simultaneously
+get_port_prefix() {
+    port_prefix="${PORT_PREFIX:-}"
+    tries=0
+    while true
+    do
+        vnc="${port_prefix}5909"
+        nc -z localhost "$vnc" || break
+        tries=$(( tries + 1 ))
+        port_prefix="${tries}"
+    done
+    echo "$port_prefix"
+}
+
+docker_image() {
     validate_python_version $1 $2
 
     docker build \
            ${DOCKER_BUILD_ARGUMENTS:-} \
            --build-arg=USER_ID="$(id -u)" \
            --build-arg=USER_NAME="${USER:-root}" \
-           -t "securedrop-test-${1}-py${2}" \
-           --file "${TOPLEVEL}/securedrop/dockerfiles/${1}/python${2}/Dockerfile" \
-           "${TOPLEVEL}/securedrop"
+           -t "sd-${1}-py${2}-$(hashpath)" \
+           --file "${REPOROOT}/securedrop/dockerfiles/${1}/python${2}/Dockerfile" \
+           "${REPOROOT}/securedrop"
 }
 
-function docker_run() {
+docker_run() {
     validate_python_version $1 $2
 
     find . \( -name '*.pyc' -o -name __pycache__ \) -delete
 
+    port_prefix="$(get_port_prefix)"
+
+    if [ -z "$port_prefix" ]
+    then
+        echo "Using no prefix for exposed Docker ports."
+    else
+        echo "************************************************************"
+        echo "Exposed Docker ports will be available on localhost prefixed"
+        echo "with \"${port_prefix}\", e.g. ${port_prefix}8080 for the source interface, and "
+        echo "${port_prefix}8081 for the journalist interface."
+        echo "************************************************************"
+    fi
+
     # The --shm-size argument sets up dedicated shared memory for the
     # container. Our tests can fail with the default of 64m.
     docker run \
            --shm-size 2g \
-           -p 127.0.0.1:5909:5909 \
            --rm \
+           -p "127.0.0.1:${port_prefix}5909:5909" \
+           -p "127.0.0.1:${port_prefix}8080:8080" \
+           -p "127.0.0.1:${port_prefix}8081:8081" \
            -e NUM_SOURCES \
            -e LC_ALL=C.UTF-8 \
            -e LANG=C.UTF-8 \
            -e PAGE_LAYOUT_LOCALES \
+           -e REPOROOT \
            --user "${USER:-root}" \
-           --volume "${TOPLEVEL}:${TOPLEVEL}" \
-           --workdir "${TOPLEVEL}/securedrop" \
-           --name securedrop-dev \
-           -ti ${DOCKER_RUN_ARGUMENTS:-} "securedrop-test-${1}-py${2}" "${@:3}"
+           --volume "${REPOROOT}:${REPOROOT}" \
+           --workdir "${REPOROOT}/securedrop" \
+           --name "sd-$(uuid)" \
```

Comment:
The executable `uuid` wasn't present on my debian-9 dev VM, it had to be installed for the second container to be triggered. Is there any reason not to use, say, $port_prefix here instead?

Reply:
Nope. It works find with the port prefix.